java/ql/src/semmle/code/java/frameworks/android/Intent.qll

```diff
@@ -27,10 +27,44 @@ class AndroidReceiveIntentMethod extends Method {
   }
 }
 
-class ContextStartActivityMethod extends Method {
-  ContextStartActivityMethod() {
-    (hasName("startActivity") or hasName("startActivities")) and
-    getDeclaringType() instanceof TypeContext
+/** A constructor of `android.content.Intent`. */
+class IntentConstructor extends TaintPreservingCallable {
+  IntentConstructor() { this.(Constructor).getDeclaringType() instanceof TypeIntent }
+
+  override predicate returnsTaintFrom(int arg) {
+    arg = 0 and this.getParameterType(arg) instanceof TypeIntent
+    or
+    arg in [1, 4] and this.getParameterType(arg) instanceof TypeClass
+  }
+}
+
+/**
+ * Methods of a subclass of `android.content.Context` that start activities.
+ */
+class StartActivityMethod extends Method {
+  StartActivityMethod() {
+    getName().matches("startActivit%") and
+    getDeclaringType().getASupertype*() instanceof TypeContext
+  }
+}
+
+/**
+ * Methods of a subclass of `android.content.Context` that start services.
+ */
+class StartServiceMethod extends Method {
+  StartServiceMethod() {
+    getName().matches("start%Service%") and
```

Comment:
Not sure if it matters much, but the trailing `%` can probably be removed. It looks like the names of all methods for starting services end with "Service".

Reply:
Nope, there seems to be `startServiceAsUser`:
https://github.com/github/codeql/pull/6397/files#diff-7ae66f265c2b252af8116598976eda1d1be0f3eb3b9088fc34964033a740b107R32

Reply:
That's a good observation: there are some calls that [are not part of the public API](https://stackoverflow.com/questions/39176648/where-is-context-startserviceasuser#39176667) but there's at least [some](https://stackoverflow.com/questions/39827724/wso2emm-rebuilding-the-mobile-agent-fails/41004248#41004248) [evidence](https://github.com/anggrayudi/android-hidden-api) suggesting people are using them anyway. That's why I initially decided to include those just in case.